src/liballoc/slice.rs

```diff
@@ -1722,6 +1722,73 @@ impl<T> [T] {
         // NB see hack module in this file
         hack::into_vec(self)
     }
+
+    /// Creates a [`Vec`] by repeating a slice `n` times.
+    ///
+    /// [`Vec`]: ../../std/vec/struct.Vec.html
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// assert_eq!([1, 2].repeat(3), vec![1, 2, 1, 2, 1, 2]);
+    /// ```
+    #[stable(feature = "repeat_slice", since = "1.27.0")]
+    pub fn repeat(&self, n: usize) -> Vec<T> where T: Copy {
+        if n == 0 {
+            return Vec::new();
+        }
+
+        // If `n` is larger than zero, it can be split as
+        // `n = 2^expn + rem (2^expn > rem, expn >= 0, rem >= 0)`.
+        // `2^expn` is the number represented by the leftmost '1' bit of `n`,
+        // and `rem` is the remaining part of `n`.
+
```

Comment:
It took me a few minutes to figure out why this function is so complex - maybe add a comment that explicitly explains that its trying to minimize the `copy_nonoverlapping` calls by always copying as large chunks as possible?

Reply:
I just moved the code around, didn't modify it. It's exactly like it was before in `str`.